### PR TITLE
Send a Zoop Feature

### LINKED ISSUE #58: Send a Zoop
- [x] Unauthenticated users do not have access to this feature
 - [x] An authenticated user should be able to send a zoop
 - [ ] Should authenticated users be able to send a Zoop to themselves?

### PR DESCRIPTION
**Pull request by dyazdani**
Closes #58 

There is now a floating action button to send a Zoop, which is only enabled if you are an authenticated user. A dialog pops up to compose the Zoop. Below are screenshots of testing.

I created a new GET all users endpoint in order to complete this feature. Here it is tested on Postman (it requires an authenticated user to access):
Screen Shot 2023-11-16 at 01 24 35

The Send Zoop button is only enabled when the user has a token from logging in and shows up on all pages:
Screen Shot 2023-11-21 at 22 51 53

The Send Zoop dialog:
Screen Shot 2023-11-21 at 22 52 22
The "Username" label on the Autocomplete field gets cut off when you click inside the element. I have an Issue to fix that:
Screen Shot 2023-11-21 at 22 53 09
Screen Shot 2023-11-21 at 22 52 31

When the Zoop is successfully sent, the dialog changes and gives you a button linked to the Zoop:
Screen Shot 2023-11-21 at 22 53 15
Screen Shot 2023-11-21 at 22 53 22

https://github.com/dyazdani/zoop/assets/99094815/5eeab7f9-331c-45e4-9225-1f5439a56b42

### CODE REVIEW

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:36]: If all you care about is getting the currently logged in user's ID for mutations, I recommend storing the user ID in the authSlice instead of fetching it every time. The User ID for a given user will never change, so I think it makes sense to store it across your app.

Also, it looks like your authSlice is already setup to support setting user, it just needs to be set in the extraReducers so you'll have access to the user from the store

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:48]: You should not need to get the user object - you should have the user stored in state, not just the username

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:28]:
```suggestion
    const [selectedUser, setSelectedUser] = useState<User | null>(null);
```

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:45]: I think this is fine - but you should also ensure that the component doesn't even render until the users list has loaded

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:62]: I'm not sure why this is necessary? You shouldn't need to add your own state management. If you care about most recent zoop sent, it should be stored in your Redux store across your whole app. And to tell whether it is sent, you should be able to check if there data is defined (it will be undefined until the mutation is complete)

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:76]: Prefer explicit text in the button to a Tooltip in cases like this. Tooltips aren't really meant for situations like this, it's a little hard to explain, but it mostly comes down to - once you know what it is, you shouldn't have to see this tooltip every time. Tooltips cause motion which is visually distracting. Save Tooltips for things like form fields with an optional tooltip someone can use to learn more about a term.

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:78]: Why show the Fab at all if there is no token? Return null from this component if token doesn't exist, the user cannot send a zoop until they log in

Comment (paloobi) [reply client/components/SendZoopDialog.tsx:78]: Other option is to never disable it, but if they click it before logging in/registering, the dialog could instead say "You must have an account to zoop!" and then shows links to the login and register pages

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:81]: personally, I prefer the top, eg in the header or just below the header, as it's more prominent.

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:91]: Don't store state of if the zoop was sent - that state is already managed by RTK Query. Just check for response data, eg. `zoopData.zoop`, is truthy. If it is truthy, the mutation has completed successfully.
```suggestion
            {zoopData.zoop ? (
```

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:101]:
```suggestion
                                onClick={() => navigate(`/zoops/${zoopData.zoop.id}`)}
```

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:103]: suggestion: keep it short and simple.

```suggestion
                                Go to Zoop.
```

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:105]: You might add a TODO to render the created zoop somehow to provide nice visual feedback. Up to you if you want to do that now vs. later

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:113]: FYI this really feels like a separate component, so you might consider separating it out to make things more readable

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:114]: suggestion: I like "Send Zoop" bc I think it's simpler? But I leave it up to you

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:125]: I would not suggest a `TextField` here - I would recommend some sort of typeahead/dropdown selection. Check out `Autocomplete` in MUI: https://mui.com/material-ui/react-autocomplete

In that case, you would use map over the users to create an array of objects, where the label is the username, and the ID is the user's ID.

You could add a Wish List task for enabling search as you type to make it more performant (this would work better if there were a lot of users): https://mui.com/material-ui/react-autocomplete/#search-as-you-type

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:132]: Could we make the label a little more fun to match the Zoop branding? eg. "Say something" or "What do you want to say?"

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:26]: This component is not a Dialog. it is a button that contains a dialog. Recommend renaming it to SendZoopButton (especially since I also recommend creating a separate SendZoopDialog component later in the review comments, and you may also want to make a separate ZoopSentDialog component as well)

File: client/features/api.tsx
Comment (paloobi) [inline client/features/api.tsx:40]: On the front-end, this is typically referred to as a create action (POST is just the name of the HTTP method). I would recommend renaming either to createZoop or, to reflect the UX better, probably sendZoop

```suggestion
      sendZoop: builder.mutation({
```

File: src/api/users.ts
Comment (paloobi) [inline src/api/users.ts:40]: Security breach! This is sending back all of the users' passwords. Remember to delete passwords before sending!!

**Review by paloobi — CHANGES_REQUESTED** [review]
I thought from our discussion it would be in the upper right corner? I think that would make it more prominent in the UI

I would also highly recommend adding the text "Zoop" to the button, because as it is a user would likely not know what it was for (maybe also test out "Send Zoop" or "Zoop It" to make it even clearer it's an action button, see what looks best)

I left a lot of comments. RC is for anything that wasn't labeled as a suggestion or a nit. Let me know if you have questions about any of the feedback

File: client/components/HomePage.tsx
Comment (dyazdani) [reply client/components/HomePage.tsx:12]: It's currently not on LoginPage or RegisterPage because at no point is the user on those pages logged in. Do you want it to be? It certainly would be DRYer to just add it to `App.tsx`. The button is disabled when the user is not logged in anyway.

Comment (paloobi) [reply client/components/HomePage.tsx:12]: Button should be hidden when not logged in. I would recommend adding it in App.tsx and just not rendering it if there is no token

**Comment by paloobi** [comment]
I do agree that the upper right is a place people look for their account/login info. I think we need new design here. During design discussions I remember that we weren't planning to have an AppBar, and that the Zoop button would be in the top right

I think the icon and text in the header can and should be much larger to match the design mock

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:48]: I was using this logic to make sure that the correct user is selected based on the username that is being asked for in the send Zoop dialog. Not sure how else to do this. How should I have done this differently?

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:28]: Is meant to be the user who is receiving the Zoop?

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:107]: I had it that way before. Maybe I'll switch back.

File: src/api/users.ts
Comment (dyazdani) [reply src/api/users.ts:40]: Thanks for the catch!

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:60]: You should also call the `onClose` from the parent - this will avoid this component remaining mounted (even if it's not visible) after it was closed

**Review by paloobi — CHANGES_REQUESTED** [review]
RC for:
- don't store isSent and sentZoop in state
- move ZoopSentDialog inside of SendZoopDialog
- make sure that your dialogs are unmounting properly when they are closed (if close button is clicked on either of them, make sure that the SendZoopButton does not render either dialog anymore)
- use onSubmit instead of onClick
- fix props in ZoopSentDialog
- use logged in user from store instead of fetching with a query

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:24]: Probably a stupid question, but why does this need to be stored in state?

Comment (paloobi) [reply client/components/SendZoopDialog.tsx:24]: Because it is a piece of data inside your form. You want to allow the person to select a user in the form and then submit a zoop to them, so you need to track which user they selected so you know which user ID to send in the sendZoop mutation

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [inline client/components/SendZoopDialog.tsx:76]: I could swear I left a comment here suggesting you take a look at the `Autocomplete` component in MUI instead of just using a text input. That would allow them to start typing a username and then see a filtered list of all usernames that match

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [reply client/components/SendZoopDialog.tsx]: @dyazdani this comment got marked as outdated, want to make sure you saw it

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx]: When you say I don't need  to "get the user" object, I take that to mean that I don't need to be calling filter on the array of users. But this was my way of getting the selected user's ID to put in the Zoop. There is no other place with the users' IDs except the user objects in this array. When you say, "the user stored in state, not just the username" are you wanting me to store a user object in state with ll of the other properties? If so, I would need to query to get it as it is not in the redux store.

Comment (dyazdani) [reply client/components/SendZoopDialog.tsx]: Although now that I'm looking at the Autocomplete component I set up, the ID in the input object (`{label: string, id: number}`) should be the same id that is in the user's object in the database because it mapped through all the users to create the objects for Autocomplete. I was thinking that the IDs wouldn't match up. I think I see what you are saying now.

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:24]: I was using the username string from the Dialog textfield to compare to the `User` objects and find the right user ID to use in the Zoop for the `receiverId`. Wouldn't storing the selected user in state be an unnecessary additional step?

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx]: You're talking about the selected user / recipient, right? What do you mean by "get the user object"? Do you mean find the selected user using a filter of `usersData`?

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:24]: But it won't be a typed `User` object, will it? The data I am getting from the Autocomplete field of the dialog is one of the objects in the array of options (each option is a type `{label: string, id: number}`). I think the `id` property on those options should match the IDs for the same users in the DB. So `selectedUser` would just be one of those option objects, right? And I could just use `selectedUser.id` as the `receiverId` on the Zoop, right? Or do you want me to find the `User` object in `usersData` and store it in the component's state?

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx]: Yes, I just finished this!

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx]: Strangely there is no `component` prop for `DialogContent` or `Dialog`. Here is what MUI has for an example of a form dialog: https://mui.com/material-ui/react-dialog/#form-dialogs.

File: client/components/ZoopSentDialog.tsx
Comment (dyazdani) [reply client/components/ZoopSentDialog.tsx:28]: Got it. I assume it would just be a snippet, i.e., not as much as the individual Zoop page for the Zoop, right?

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx]: Yeah, I got that comment. It's done!

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [reply client/components/SendZoopDialog.tsx]: I meant you shouldn't need to filter to find which user was selected after they select it (that means looping through the entire list of all users before you can submit), you should instead select the user in state by ID with the Autocomplete, and then you will have that user's ID stored in state.

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [reply client/components/SendZoopDialog.tsx]: Yes, the ID in Autocomplete should match the ID of the user in the DB. So selectedUser should be whatever Autocomplete expects (can't remember if the selected value is the `{label: String, id: ID}` object or if it's just the ID). And then you will have the ID saved in your state so you can reference it to make the mutation.

The docs for Autocomplete should have some guidance for how to store the state correctly, from the examples it looks like it's probably a new object that has label & id - but either way, that means you can extract the ID from state: https://mui.com/material-ui/react-autocomplete

Comment (dyazdani) [reply client/components/SendZoopDialog.tsx]: Ahhhhh, I didn't realize you wanted me to make a new mutation. Can do. Makes sense.

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [reply client/components/SendZoopDialog.tsx:77]: Take a look at the controlled state examples: https://mui.com/material-ui/react-autocomplete/#controlled-states (remember to expand to see the code)

In your case, you may not need to control the input value state, just the selected value.

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [reply client/components/SendZoopDialog.tsx:95]: @dyazdani - this still needs to be changed to a form with onSubmit

**Review by paloobi — CHANGES_REQUESTED** [review]
RC for:
- finish fixes for Autocomplete (eg. make it controlled, stop filtering through list of all users by username, and make sure the options are setup correctly)
- use a form with onSubmit instead of onClick

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:36]: What would handling it more gracefully be? Displaying something in the UI?

**Comment by dyazdani** [comment]
@paloobi MUI doesn't have a form component so I've tried using `FormControl`, but it doesn't seem to do anything on submit. I am not quite sure I am using it right. I followed this tutorial: https://www.copycat.dev/blog/material-ui-form/

Should I be using a regular form element instead?

Also, just curious, does it need to be a form for semantic purposes?

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:95]: @paloobi MUI doesn't have a form component so I've tried using `FormControl`, but it doesn't seem to do anything on submit. I am not quite sure I am using it right. I followed this tutorial: https://www.copycat.dev/blog/material-ui-form/

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [reply client/components/SendZoopDialog.tsx]: There shouldn't need to be a new mutation other than `sendZoop`? You should be able to call `sendZoop` with the selected user ID

File: client/components/SendZoopDialog.tsx
Comment (paloobi) [reply client/components/SendZoopDialog.tsx:36]: Displaying something in the UI is great - but if you don't have time for that, you could put a console.error instead of throwing

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:62]: The objects do have a user property (that property is assigned the return value of an `excludePassword(user)` call in the endpoint code you cited here). 

The returned response is: 
```
{
  users: [
    {user: {id: ..., username: ...}}
    {user: {id: ..., username: ...}}
    {user: {id: ..., username: ...}}
  ]
}
```

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:77]: Good call. I got distracted by the value argument in the onChange callback, thinking that would give me the value just fine, but you are right that I wasn't using state and it's probably not actually controlling the component.

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:75]: I added the `isOptionEqualToValue` to deal with an error that cropped up. Let's see if I need that if I control the component.

Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:75]: It was a console warning I got: 
Screen Shot 2023-11-20 at 15 02 28

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:62]: Also, when I used the `useMemo` call you suggested, the following error appears and nothing is rendered after clicking the Send Zoop button (blank screen):
Screen Shot 2023-11-20 at 18 12 58

I can try to troubleshoot later tonight or tomorrow.

**Comment by dyazdani** [comment]
Made all changes except for the error with `useMemo`. See my comment above.

File: client/components/ZoopSentDialog.tsx
Comment (paloobi) [reply client/components/ZoopSentDialog.tsx:28]: Probably just the receiver and contents. It's good to confirm after an action what the user just did

File: client/components/ZoopSentDialog.tsx
Comment (paloobi) [inline client/components/ZoopSentDialog.tsx:28]: Don't recommend using a Link with navigate.

Prefer using React Router link (using component prop). Check out the example in Robot Pet City to see how to make TS happy

**Review by paloobi — CHANGES_REQUESTED** [review]
RC for:

- useMemo
- use react router link
- add screen recordings, and also update screenshots in PR description (if you have trouble getting screen recordings small enough file size, you can export them with lower quality from QuickTime)
-

File: client/components/SendZoopDialog.tsx
Comment (dyazdani) [reply client/components/SendZoopDialog.tsx:81]: Moving useMemo worked! Thanks!

File: client/components/ZoopSentDialog.tsx
Comment (dyazdani) [reply client/components/ZoopSentDialog.tsx:28]: Out of curiosity, is there a reason to use React Router link over MUI's Link with useNavigate?

Comment (paloobi) [reply client/components/ZoopSentDialog.tsx:28]: In my experience when a link has an onClick most libraries no longer treat it as a link but as a button. So HTML would not be semantic.

Also HTML link has a URL location, which is what React Router link will create for you. So it is more easy for users to tell where the link will go. It will behave like a link (eg. Showing the link in bottom of browser window while hovering, or maybe reading where the URL will be for some screen readers depending on settings).

So those are my instincts, since MUI has no idea what React Router is doing (eg. What navigate function is) I believe this leads to less semantic and less accessible markup.

**Review by paloobi — APPROVED** [review]
Approving to unblock, but recommend either adding react router link that renders as MUI link, or at least filing an issue and adding a to-do to track it before merging.

**Comment by dyazdani** [comment]
> Approving to unblock, but recommend either adding react router link that renders as MUI link, or at least filing an issue and adding a to-do to track it before merging.

Would that be nesting a MUI Link in a React Router Link?

**Pull request merged**
